pkg/pool-weighted/contracts/smart/IndexPool.sol

```diff
@@ -95,22 +129,146 @@ contract IndexPool is BaseWeightedPool {
+                endWeights[i] =  _tokenState[_tokens[i]].decodeUint32(_END_WEIGHT_OFFSET).uncompress32();
+            }
+    }
+
+    /**
+     * @dev Returns a fixed-point number representing how far along the current weight change is, where 0 means the
+     * change has not yet started, and FixedPoint.ONE means it has fully completed.
+     */
+    function _calculateWeightChangeProgress(bytes32 poolState) private view returns (uint256) {
+        uint256 currentTime = block.timestamp;
+        uint256 startTime = poolState.decodeUint32(_START_TIME_OFFSET);
+        uint256 endTime = poolState.decodeUint32(_END_TIME_OFFSET);
+
+        if (currentTime > endTime) {
+            return FixedPoint.ONE;
+        } else if (currentTime < startTime) {
+            return 0;
+        }
+
+        // No need for SafeMath as it was checked right above: endTime >= currentTime >= startTime
+        uint256 totalSeconds = endTime - startTime;
+        uint256 secondsElapsed = currentTime - startTime;
+
+        // In the degenerate case of a zero duration change, consider it completed (and avoid division by zero)
+        return totalSeconds == 0 ? FixedPoint.ONE : secondsElapsed.divDown(totalSeconds);
+    }
+
+    /**
+    * @dev When calling updateWeightsGradually again during an update, reset the start weights to the current weights,
+    * if necessary.
+    */
+    function _startGradualWeightChange(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory startWeights,
+        uint256[] memory endWeights
+    ) internal virtual {
+        bytes32 newPoolState = _poolState;
+
+        uint256 normalizedSum = 0;
+        for (uint256 i = 0; i < endWeights.length; i++) {
+            uint256 endWeight = endWeights[i];
+            _require(endWeight >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
+
+            _tokenState[_tokens[i]] = _tokenState[_tokens[i]]
+            .insertUint64(startWeights[i].compress64(), _START_WEIGHT_OFFSET )
+            .insertUint32(endWeight.compress32(), _END_WEIGHT_OFFSET)
+            .insertUint5(uint256(18).sub(ERC20(address(_tokens[i])).decimals()), _DECIMAL_DIFF_OFFSET);
+
+            normalizedSum = normalizedSum.add(endWeight);
+        }
+        // Ensure that the normalized weights sum to ONE
+        _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
+
+        _poolState = newPoolState.insertUint32(startTime, _START_TIME_OFFSET).insertUint32(endTime, _END_TIME_OFFSET);
+
+        //        emit GradualWeightUpdateScheduled(startTime, endTime, startWeights, endWeights);
+    }
+
+
+    /**
+     * @dev Schedule a gradual weight change, from the current weights to the given endWeights,
+     * over startTime to endTime
+     */
+//    TODO: has to be authenticate
+    function updateWeightsGradually(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory endWeights
+    ) internal whenNotPaused nonReentrant {
```

Comment:
Do we want to add the function to pause for the index pool, because we have the modifier `whenNotPaused` in this function? I didn't come across a pause possibility in the V1 indexPool contract.

Reply:
Hmm, nice question, I don't think that we need to, let's add it when we'll be asked to

Reply:
Done